internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -677,12 +702,41 @@ func (r *ClusterDeploymentsReconciler) deregisterClusterIfNeeded(ctx context.Con
 	}); err != nil {
 		return buildReply(err)
 	}
+	// Delete agents because their backend cluster got deregistered.
+	if err = r.DeleteClusterDeploymentAgents(ctx, key); err != nil {
+		return buildReply(err)
+	}
 
 	r.Log.Infof("Cluster resource deleted, Unregistered cluster: %s", c.ID.String())
 
 	return buildReply(nil)
 }
 
+func (r *ClusterDeploymentsReconciler) DeleteClusterDeploymentAgents(ctx context.Context, clusterDeployment types.NamespacedName) error {
+	agents := &aiv1beta1.AgentList{}
+	if err := r.List(context.Background(), agents); err != nil {
+		return err
+	}
+	for _, clusterAgent := range agents.Items {
+		if clusterAgent.Spec.ClusterDeploymentName.Name == clusterDeployment.Name &&
+			clusterAgent.Spec.ClusterDeploymentName.Namespace == clusterDeployment.Namespace {
+			r.Log.Infof("clusterDeployment %s namespace %s, delete agent %s",
+				clusterDeployment.Name, clusterDeployment.Namespace, clusterAgent.Name)
+			agentToDelete := &aiv1beta1.Agent{
```

Comment:
Can you just use `clusterAgent` instead of instantiating a new Agent?

Reply:
Done